### ISSUE TITLE
C parser can't handle structure typedefs referencing themselves as a pointer (such as linked lists)

### ISSUE THREAD

**Issue by baconwaifu** [NationalSecurityAgency/ghidra#3505]
**Describe the bug**
A structure that refers to other instances via pointer causes the C parser to fail, as the type is not defined. This breaks a lot of linked-list implementations if the list nodes are `typedef`ed.

**To Reproduce**
```c
typedef struct Il2CppStringBuilder
{
    Il2CppObject object;
    Il2CppArray* chunkChars;
    Il2CppStringBuilder* chunkPrevious; // This line will fail as Il2CppStringBuilder is not defined yet
    //struct Il2CppStringBuilder* chunkPrevious; // And this is the workaround
    int chunkLength;
    int chunkOffset;
    int maxCapacity;
} Il2CppStringBuilder; // Type 'learned' here
```

**Expected behavior**
The parser doesn't fall over on a basic linked-list

**Environment (please complete the following information):**
 - OS: Debian bullseye
 - Java Version: openjdk-11
 - Ghidra Version: 10.0.4
 - Ghidra Origin: Github Release

**Additional context**
Possible fix is to defer the type lookup/validity check for unknown pointer types until the rest of the file/block is parsed? Shouldn't complicate the parser too much.

A workaround is to use `struct` to refer to the self-referencing type, as ghidra already knows that type exists by the time it parses the line in question.

**Comment by quietust** [comment]
From what I recall, C does not allow a typedef to be used inside itself but instead requires you to mention the `struct`. I'm not sure about the rules in C++, though.

**Comment by astrelsky** [comment]
In c either a forward declaration or the use of the struct keyword is required. I don't think it is required in c++ though.

**Comment by mid-kid** [comment]
Forward declarations don't seem to work here either...

**Comment by emteere** [comment]
It can be defined earlier in the parsing stream with a placeholder. I have a prototype fix.

**Comment by ryanmkurtz** [comment]
Fixed by f4e9dbb13a85cea0dc4af176ed72eb490db5fed4
